src/daft-logical-plan/src/ops/source.rs

```diff
@@ -13,28 +18,103 @@ pub struct Source {
 
     /// Information about the source data location.
     pub source_info: Arc<SourceInfo>,
+    pub stats_state: StatsState,
 }
 
 impl Source {
     pub fn new(output_schema: SchemaRef, source_info: Arc<SourceInfo>) -> Self {
         Self {
             output_schema,
             source_info,
+            stats_state: StatsState::NotMaterialized,
         }
     }
 
+    // Helper method that converts the ScanOperatorRef inside a Source node's PhysicalScanInfo into scan tasks.
+    // Should only be called if a Source node's source info contains PhysicalScanInfo. The PhysicalScanInfo
+    // should also hold a ScanState::Operator and not a ScanState::Tasks (which would indicate that we're
+    // materializing this physical scan node multiple times).
+    pub(crate) fn build_materialized_scan_source(
+        mut self,
+        execution_config: Option<&DaftExecutionConfig>,
+    ) -> DaftResult<Self> {
+        let new_physical_scan_info = match Arc::unwrap_or_clone(self.source_info) {
+            SourceInfo::Physical(mut physical_scan_info) => {
+                let scan_tasks = match &physical_scan_info.scan_state {
+                    ScanState::Operator(scan_op) => scan_op
+                        .0
+                        .to_scan_tasks(physical_scan_info.pushdowns.clone(), execution_config)?,
```

Comment:
I think we should avoid execution_config completely in the logical planner. This would also mean that the logical plan needs to know what execution engine we are targeting which breaks the invariant of the logical plan.

Right now `to_scan_tasks` performs generates the scan tasks and then also run the scan task passes. But the implementation of that looks identical for the 3 ScanOperator impls that we have.

My proposal is that we remove the scan task passes from the  `to_scan_tasks` for `GlobScanOperator`, `AnonymousScanOperator` and the `PythonScanOperator` and then run it during translation.

Reply:
Sounds good!

I did this, with a few caveats:
- At translation time, we have already converted the concrete scan tasks into `Arc<Vec<Arc<dyn ScanTaskLike>>>`
- I implemented `ScanTaskLike.split()` as a trait method, then modified `split_by_row_groups` to use this trait method on our trait objects
- Unfortunately I could not do the same for merging because merging requires some data source and partitioning information that, if pulled into `ScanTaskLike` would create circular dependencies.
- Instead, since the only non-test implementer of `ScanTaskLike` is `ScanTask`, for merging I downcasted the `Arc<dyn ScanTaskLike>`s into `Arc<ScanTask>`, then kept the old implementation of merging.

Please let me know if this works for you or if it's a little sus!

Reply:
Left some feedback on this! Happy to pair up tomorrow to crank this out